doc/adr/2019-02-13-preserve-all-functionality.md

```diff
@@ -0,0 +1,31 @@
+**Title**: Preserve all functionality offered by the raw API.
+
+**Status**: Proposed
+
+**Context**: When we design APIs for the client libraries one may need to
+tradeoff ease of use for functionality. All things being equal it is easier to
+design simpler, easier to understand client libraries if one is willing to
+remove functionality that is rarely used. While we want to encourage users to
+adopt the libraries, we do not want to abandon advanced users to their own
+devices. We do not want them to be tempted or forced to discard the client
+libraries in favor of raw REST or gRPC because that is the only way they can
+access a feature.
+
+**Decision**: The libraries must preserve all functionality. If it is possible
```

Comment:
What do we think about allowing users to access (or giving them access) to the underlying gapic-generated libraries when/if they need the full power of some rarely used part of the API? That would enable the advanced/rare use cases by letting them directly interact w/ the grpc or gapic libraries. But perhaps then our higher-level C++ libraries could focus on making the 95% use cases really nice and easy?

FWIW: I'm not sure of my preference/opinion on this point. Just raising the question to see what others think.

Reply:
I am in favor of telling users where to find the GAPIC libraries.  Pragmatically, it is not like we can stop users from just compiling the gRPC and/or GAPIC libraries themselves if they really want to. In some cases I am in favor of pointing them directly to GAPIC (e.g. the admin APIs rarely get much better by doing a lot of hand-crafted code).

I think the issue will be (as usual) on how much removed functionality is "too much", or where exactly is that 95%, or if a particular use-case is in the 95% or not.

I am feeling more and more inclined to say: document these cases clearly **and** create ~bugs~feature-requests on GitHub so we can get votes and tracking for them.